Mateo/dev 25 write connecting arcade tools to your llm page (#595)
ArcadeAI/docs: app/en/home/connect-arcade-to-your-llm/page.mdx

---
title: "Connect Arcade to your LLM"
description: "Learn how to connect Arcade to your LLM"
---

import { Steps, Tabs, Callout } from "nextra/components";
import { SignupLink } from "@/app/_components/analytics";

# Connect Arcade to your LLM

<GuideOverview>
<GuideOverview.Outcomes>

> Integrate Arcade tool-calling capabilities into an application that uses an LLM in Python.

Reviewer: Defer to @vfanelle on this phrasing

- The [`uv` package manager](https://docs.astral.sh/uv/getting-started/installation/)

</GuideOverview.Prerequisites>

<GuideOverview.YouWillLearn>

- Setup a simple agentic loop
- Add Arcade tools to your agentic loop
- Implement a multi-turn conversation loop

</GuideOverview.YouWillLearn>
</GuideOverview>

<Steps>

### Create a new project and install the dependencies

In your terminal, run the following command to create a new `uv` project

```bash
mkdir arcade-llm-example
cd arcade-llm-example
uv init
```

Create a new virtual environment and activate it:

```bash
uv venv
source .venv/bin/activate
```

Install the dependencies:

```bash
uv add arcadepy openai python-dotenv
```

Your directory should now look like this:

```bash
arcade-llm-example/
├── .git/
├── .gitignore
├── python-version
├── .venv/
├── main.py
├── pyproject.toml
├── main.py
├── README.md
└── uv.lock
```

### Instantiate and use the clients

Create a new file called `.env` and add your Arcade API key, as well as your OpenAI API key:

```bash
ARCADE_API_KEY=YOUR_ARCADE_API_KEY
ARCADE_USER_ID=YOUR_ARCADE_USER_ID
OPENROUTER_API_KEY=YOUR_OPENROUTER_API_KEY
OPENROUTER_MODEL=YOUR_OPENROUTER_MODEL
```

> 

Reviewer: You might want to remind folks where they can get their Arcade User ID

  OpenRouter is compliant with the OpenAI API specification, so you can use it
  with any OpenAI-compatible library.
</Callout>

Open the `main.py` file in your editor of choice, and replace replace the contents with the following:

```python filename="main.py"
from arcadepy import Arcade
from openai import OpenAI
from dotenv import load_dotenv
import json
import os

load_dotenv()

client = Arcade()
arcade_user_id = os.getenv("ARCADE_USER_ID")
llm_client = OpenAI(
    api_key=os.getenv("OPENROUTER_API_KEY"),
    base_url="https://openrouter.ai/api/v1"
)

```

### Select and retrieve the tools from Arcade

In this example, we're implementing a simple agent that can retrieve and send emails, as well as send messages to Slack. In a harness like this one, where the entire catalog of tools is exposed to the LLM, you will want to choose only the tools that are relevant to the task at hand to avoid overwhelming the LLM with too many options, and to make your agent more token efficient.

```python filename="main.py"
# We define here the tools that we want to use in the agent
tool_catalog = [
    "Gmail.ListEmails",
    "Gmail.SendEmail",
    "Slack.SendMessage",
    "Slack.WhoAmI"
]

> # We get the tool definitions from the Arcade API, so that we can expose them to the LLM

Reviewer: ```suggestion
# Get the tool definitions from the Arcade API 
```

tool_definitions = []
for tool in tool_catalog:
    tool_definitions.append(arcade.tools.formatted.get(name=tool, format="openai"))
```

### Write a helper function that handles tool authorization and execution

The LLM may choose to call any of the tools in our catalog, and we need to handle different scenarios, such as when the tool requires authorization. There are many approaches to this, but a good pattern is to handle interruptions like this outside of the agentic context if we can do so. As a rule of thumb, you should evaluate whether it is relevant for the LLM to be aware of the authorization process, or if it's better to handle it in the harness, and keep the context as if the tool was already authorized. The latter option optimizes for token efficiency, and we will implement it in this example.

> ```python filename="main.py"

Reviewer: ```suggestion
```python filename="hello.py"
```

                tool_result = authorize_and_run_tool(tool_name, tool_args)
                print(f"🛠️ Harness: Tool call {tool_name} completed")
                history.append({
                    "role": "tool",
                    "tool_call_id": tool_call.id,
                    "content": tool_result,
                })

            continue

        else:
            history.append({
                "role": "assistant",
                "content": assistant_message.content,
            })

        break

    return history
```

In combination, these two helper functions will form the core of our agentic loop. You will notice that the authorization is handled outside of the agentic context, and the tool execution is passed back to the LLM in every case. Depending on your needs, you may want to handle tool orchestration within the harness, and pass only the final result of multiple tool calls to the LLM.

### Write the main agentic loop

Now that we've written the helper functions, we can write a very simple agentic loop that interacts with the user. The core pieces of this loop are:

1. Initialize the conversation history with the system prompt

> 2. Get the user input, and add it to the conversation history

Reviewer: ```suggestion
2. Get the user input and add it to the conversation history
```

- "Please write a poem about multi-tool orchestration and send it to the #general channel on Slack, also send it to me in an email."
- "Please summarize my latest 5 emails, then send me a DM on Slack with the summary."

</Steps>

## Next Steps

- Learn more about using Arcade with frameworks like [LangChain](/home/langchain/use-arcade-tools) or [Mastra](/home/mastra/overview).
- Learn more about how to [build your own MCP Servers](/home/build-tools/create-a-mcp-server).

## Example code

```python filename="main.py"
from arcadepy import Arcade
from dotenv import load_dotenv
from openai import OpenAI
import json
import os

load_dotenv()

arcade = Arcade()
arcade_user_id = os.getenv("ARCADE_USER_ID")
llm_client = OpenAI(
    base_url="https://openrouter.ai/api/v1",
    api_key=os.getenv("OPENROUTER_API_KEY"),
)

# We define here the tools that we want to use in the agent
tool_catalog = [
    "Gmail.ListEmails",
    "Gmail.SendEmail",
    "Slack.SendMessage",
    "Slack.WhoAmI"
]

# We get the tool definitions from the Arcade API, so that we can expose them to the LLM
tool_definitions = []
for tool in tool_catalog:

>     tool_definitions.append(arcade.tools.formatted.get(name=tool, format="openai"))

Reviewer: But the code errors out `arcade`. If it don't run, we must fix. 

Please choose a different variable name than `client` and update the code. Thanks!

- The [`uv` package manager](https://docs.astral.sh/uv/getting-started/installation/)

</GuideOverview.Prerequisites>

<GuideOverview.YouWillLearn>

- Setup a simple agentic loop
- Add Arcade tools to your agentic loop
- Implement a multi-turn conversation loop

</GuideOverview.YouWillLearn>
</GuideOverview>

<Steps>

### Create a new project and install the dependencies

In your terminal, run the following command to create a new `uv` project

```bash
mkdir arcade-llm-example
cd arcade-llm-example
uv init
```

Create a new virtual environment and activate it:

```bash
uv venv
source .venv/bin/activate
```

Install the dependencies:

```bash
uv add arcadepy openai python-dotenv
```

Your directory should now look like this:

```bash
arcade-llm-example/
├── .git/
├── .gitignore
├── python-version
├── .venv/
├── main.py
├── pyproject.toml
├── main.py
├── README.md
└── uv.lock
```

### Instantiate and use the clients

Create a new file called `.env` and add your Arcade API key, as well as your OpenAI API key:

```text filename=".env"
ARCADE_API_KEY=YOUR_ARCADE_API_KEY
ARCADE_USER_ID=YOUR_ARCADE_USER_ID
OPENROUTER_API_KEY=YOUR_OPENROUTER_API_KEY

> OPENROUTER_MODEL=YOUR_OPENROUTER_MODEL

Reviewer 2: Let's recommend a model to use. gpt-5.2?

```

<Callout type="info">
  The `ARCADE_USER_ID` is the email address you used to sign up for Arcade. When
  your app is ready for production, you can set this dynamically based on your
  app's auth system. Learn more about how to achieve secure auth in production
  [here](/home/auth/secure-auth-production).
</Callout>

<Callout type="info">
  In this example, we're using OpenRouter to access the model, as it makes it
  very easy to use any model from multiple providers with a single API.

OpenRouter is compliant with the OpenAI API specification, so you can use it
with any OpenAI-compatible library.

</Callout>

> Open the `main.py` file in your editor of choice, and replace replace the contents with the following:

Reviewer 2: ```suggestion
Open the `main.py` file in your editor of choice. Find and replace the contents with the following:
```

```python filename="main.py"
from arcadepy import Arcade
from openai import OpenAI
from dotenv import load_dotenv
import json
import os

load_dotenv()

> client = Arcade()

Reviewer 2: Fixed a variable naming bug that prevented the code snippets from executing.

arcade_user_id = os.getenv("ARCADE_USER_ID")
llm_client = OpenAI(
    api_key=os.getenv("OPENROUTER_API_KEY"),
    base_url="https://openrouter.ai/api/v1"
)

```

### Select and retrieve the tools from Arcade

> In this example, we're implementing a multi-tool agent that can retrieve and send emails, as well as send messages to Slack. In a harness like this one the entire catalog of tools is exposed to the LLM. To make your agent more efficient, choose only the tools that are relevant to the task at hand to avoid overwhelming the LLM and to make your agent more token efficient.

Reviewer 2: ```suggestion
In this example, we're implementing a multi-tool agent that can retrieve and send emails, as well as send messages to Slack. In a harness like this one, the entire catalog of tools is exposed to the LLM. To make your agent more efficient, choose only the tools that are relevant to the task at hand to avoid overwhelming the LLM and to make your agent more token efficient.
```

> In this example, we're implementing a multi-tool agent that can retrieve and send emails, as well as send messages to Slack. In a harness like this one the entire catalog of tools is exposed to the LLM. To make your agent more efficient, choose only the tools that are relevant to the task at hand to avoid overwhelming the LLM and to make your agent more token efficient.

Reviewer 2: ```suggestion
In this example, we're implementing a multi-tool agent that can retrieve and send emails, as well as send messages to Slack. While a harness can expose a broad catalog of tools to the LLM, it’s best to limit that set to what’s relevant for the task to keep the model efficient.
```

```python filename="main.py"
# Define the tools for the agent to use
tool_catalog = [
    "Gmail.ListEmails",
    "Gmail.SendEmail",
    "Slack.SendMessage",
    "Slack.WhoAmI"
]

# Get the tool definitions from the Arcade API
tool_definitions = []
for tool in tool_catalog:
    tool_definitions.append(arcade.tools.formatted.get(name=tool, format="openai"))
```

### Write a helper function that handles tool authorization and execution

> The LLM may choose to call any of the tools in its catalog, some of which may require authorization, interrupting the process.

Reviewer 2: ```suggestion
The model can use any tool you give it, and some tools require permission before they work. When this happens, you can either involve the model in the permission step or handle it behind the scenes and continue as if the tool were already authorized. In this guide, authorization is handled outside the model so it can act as if the tool is already available. It’s like ordering a coffee: after you place your order, the barista handles payment behind the counter instead of explaining every step of card verification and receipts. The customer—and the model—gets the result without having to think about any of the intermediate steps.
```
